src/main/java/io/bf2/kafka/authorizer/CustomAclAuthorizer.java

```diff
@@ -421,28 +422,23 @@ public List<CompletionStage<AclCreateResult>> createAcls(AuthorizableRequestCont
             .map(binding -> {
                 final CompletionStage<AclCreateResult> result;
 
-                if (!binding.entry().principal().startsWith(CustomAclBinding.USER_TYPE_PREFIX)) {
-                    /* Reject ACL operations as invalid where the principal named in the ACL binding is the principal performing the operation */
-                    log.info("Rejected attempt by user {} to create ACL binding with invalid principal name: {}",
+                String bindingPrincipal = binding.entry().principal();
+                if (!bindingPrincipal.startsWith(CustomAclBinding.USER_TYPE_PREFIX)) {
+                    log.info("Rejected attempt by user {} to create ACL binding with incorrect prefixing: {}",
                             requestContext.principal().getName(),
-                            binding.entry().principal());
-                    result = errorResult(AclCreateResult::new, CREATE_ACL_INVALID_PRINCIPAL);
-                } else if (Objects.equals(toString(requestContext.principal()), binding.entry().principal())) {
-                    /* Reject ACL operations as invalid where the principal named in the ACL binding is the principal performing the operation */
-                    log.info("Rejected attempt by user {} to self-assign ACL binding",
-                            requestContext.principal().getName());
-                    result = errorResult(AclCreateResult::new, CREATE_ACL_INVALID_PRINCIPAL);
-                } else if (hasPrincipalBindings(binding.entry().principal())) {
-                    /* Reject ACL operations as invalid where the principal named in the ACL binding is a principal with configured custom ACLs */
-                    log.info("Rejected attempt by user {} to create ACL binding for principal {} with existing custom ACL configuration",
+                            bindingPrincipal);
+                    result = errorResult(AclCreateResult::new, INVALID_ACL_PRINCIPAL_NON_USER_PREFIXED);
+                } else if (hasPrincipalBindings(bindingPrincipal)) {
+                    /* Reject ACL operations as invalid where the principal named in the ACL binding is a restricted principal with configured custom ACLs */
+                    log.info("Rejected attempt by user {} to create ACL binding for restricted principal {}",
                             requestContext.principal().getName(),
-                            binding.entry().principal());
-                    result = errorResult(AclCreateResult::new, CREATE_ACL_INVALID_PRINCIPAL);
+                            bindingPrincipal);
+                    result = errorResult(AclCreateResult::new, INVALID_ACL_PRINCIPAL_RESTRICTED);
```

Comment:
I was wondering whether the "owner" case should be be handled separately, with the specific error message making it clear that the creation of ACL rules with the owner's principal are disallowed.

Reply:
Sounds good, but we could defer it to another PR. Maybe it would make more sense bundled up with the owner rule visibility changes as "owner rule improvements"

If we are accumulating multiple behaviours in the plugin around the Owners then I would back off my last suggestion of having per-rule tags. Making the plugin aware of the owners starts to make more sense.